All master-eligible nodes may be upgraded before all master-ineligible nodes

To reproduce:
1. Create the following deployment:

```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: elasticsearch-sample
spec:
  version: 7.16.1
  updateStrategy:
    changeBudget:
      maxUnavailable: 2
  nodeSets:
  - name: content
    config:
      node.roles: ["data_content"]
      node.store.allow_mmap: false
    count: 2
  - name: master
    config:
      node.roles: ["master"]
      node.store.allow_mmap: false
    count: 1
```

2. Create an index with the following settings:

```json
{
  "settings": {
    "index": {
      "number_of_shards": 2,
      "number_of_replicas": 1
    }
  }
}
```

3. Upgrade the deployment (to `7.16.2` for example)

The master node instance is upgraded/deleted before one of the two data node, which is not able to join the cluster anymore:

```
"stacktrace": ["org.elasticsearch.transport.RemoteTransportException: [elasticsearch-sample-es-master-0][10.48.178.44:9300][internal:cluster/coordination/join]",
"Caused by: java.lang.IllegalStateException: node version [7.16.1] may not join a cluster comprising only nodes of version [7.16.2] or greater",
```

The reason is that the predicate [`do_not_delete_last_master_if_data_nodes_are_not_upgraded`](https://github.com/elastic/cloud-on-k8s/blob/f2fcf2eb474f9dd409d3724a85a1248e02a3c1ce/pkg/controller/elasticsearch/driver/upgrade_predicates.go#L355-L386) only relies on the label `"elasticsearch.k8s.elastic.co/node-data"` to assess if a node is not a master node: https://github.com/elastic/cloud-on-k8s/blob/f2fcf2eb474f9dd409d3724a85a1248e02a3c1ce/pkg/controller/elasticsearch/driver/upgrade_predicates.go#L380

--- a/pkg/controller/elasticsearch/driver/fixtures.go
+++ b/pkg/controller/elasticsearch/driver/fixtures.go
@@ -12,6 +12,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/uuid"
 
 	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/version"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
@@ -25,12 +26,13 @@ const (
 )
 
 type testPod struct {
-	name                                                     string
-	version                                                  string
-	ssetName                                                 string
-	master, data, healthy, toUpgrade, inCluster, terminating bool
-	uid                                                      types.UID
-	resourceVersion                                          string
+	name                                       string
+	version                                    string
+	ssetName                                   string
+	roles                                      []string
+	healthy, toUpgrade, inCluster, terminating bool
+	uid                                        types.UID
+	resourceVersion                            string
 }
 
 func newTestPod(name string) testPod {
@@ -41,8 +43,6 @@ func newTestPod(name string) testPod {
 	}
 }
 
-func (t testPod) isMaster(v bool) testPod               { t.master = v; return t }
-func (t testPod) isData(v bool) testPod                 { t.data = v; return t }
 func (t testPod) isInCluster(v bool) testPod            { t.inCluster = v; return t }
 func (t testPod) isHealthy(v bool) testPod              { t.healthy = v; return t }
 func (t testPod) needsUpgrade(v bool) testPod           { t.toUpgrade = v; return t }
@@ -50,6 +50,13 @@ func (t testPod) isTerminating(v bool) testPod          { t.terminating = v; ret
 func (t testPod) withVersion(v string) testPod          { t.version = v; return t }
 func (t testPod) inStatefulset(ssetName string) testPod { t.ssetName = ssetName; return t }
 func (t testPod) withResourceVersion(rv string) testPod { t.resourceVersion = rv; return t } //nolint:unparam
+func (t testPod) withRoles(roles ...esv1.NodeRole) testPod {
+	t.roles = make([]string, len(roles))
+	for i := range roles {
+		t.roles[i] = string(roles[i])
+	}
+	return t
+}
 
 // filter to simulate a Pod that has been removed while upgrading
 // unfortunately fake client does not support predicate
@@ -253,13 +260,23 @@ func (t testPod) toPod() corev1.Pod {
 			ResourceVersion:   t.resourceVersion,
 		},
 	}
-	labels := map[string]string{}
-	labels[label.VersionLabelName] = t.version
-	labels[label.ClusterNameLabelName] = TestEsName
-	label.NodeTypesMasterLabelName.Set(t.master, labels)
-	label.NodeTypesDataLabelName.Set(t.data, labels)
-	labels[label.StatefulSetNameLabelName] = t.ssetName
-	pod.Labels = labels
+
+	if t.version == "" {
+		t.version = "7.4.0"
+	}
+	pod.Labels = label.NewPodLabels(
+		types.NamespacedName{
+			Namespace: TestEsNamespace,
+			Name:      TestEsName,
+		},
+		t.ssetName,
+		version.MustParse(t.version),
+		&esv1.Node{
+			Roles: t.roles,
+		},
+		"https",
+	)
+
 	if t.healthy {
 		pod.Status = corev1.PodStatus{
 			Conditions: []corev1.PodCondition{

--- a/pkg/controller/elasticsearch/driver/upgrade_predicates.go
+++ b/pkg/controller/elasticsearch/driver/upgrade_predicates.go
@@ -351,7 +351,7 @@ var predicates = [...]Predicate{
 		},
 	},
 	{
-		// Force an upgrade of all the data nodes before upgrading the last master
+		// Force an upgrade of all the master-ineligible nodes before upgrading the last master-eligible
 		name: "do_not_delete_last_master_if_data_nodes_are_not_upgraded",
 		fn: func(
 			context PredicateContext,
@@ -372,13 +372,13 @@ var predicates = [...]Predicate{
 					return true, nil
 				}
 			}
-			// This is the last master, check if all data nodes are up to date
+			// This is the last master, check if all master-ineligible nodes are up to date
 			for _, pod := range context.toUpdate {
 				if candidate.Name == pod.Name {
 					continue
 				}
-				if label.IsDataNode(pod) {
-					// There's still a data node to update
+				if !label.IsMasterNode(pod) {
+					// There's still some master-ineligible nodes to update
 					return false, nil
 				}
 			}

--- a/pkg/controller/elasticsearch/driver/upgrade_predicates.go
+++ b/pkg/controller/elasticsearch/driver/upgrade_predicates.go
@@ -352,7 +352,7 @@ var predicates = [...]Predicate{
 	},
 	{
 		// Force an upgrade of all the master-ineligible nodes before upgrading the last master-eligible
-		name: "do_not_delete_last_master_if_data_nodes_are_not_upgraded",
+		name: "do_not_delete_last_master_if_all_master_ineligible_nodes_are_not_upgraded",
 		fn: func(
 			context PredicateContext,
 			candidate corev1.Pod,

--- a/pkg/controller/elasticsearch/driver/upgrade_predicates_test.go
+++ b/pkg/controller/elasticsearch/driver/upgrade_predicates_test.go
@@ -52,8 +52,8 @@ func TestUpgradePodsDeletion_WithNodeTypeMutations(t *testing.T) {
 			fields: fields{
 				esVersion: "7.2.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masterdata-0").withVersion("7.2.0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true).inStatefulset("masterdata"),
-					newTestPod("other-master-0").withVersion("7.2.0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).inStatefulset("other-master"),
+					newTestPod("masterdata-0").withVersion("7.2.0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true).inStatefulset("masterdata"),
+					newTestPod("other-master-0").withVersion("7.2.0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).inStatefulset("other-master"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchGreenHealth},
@@ -69,8 +69,8 @@ func TestUpgradePodsDeletion_WithNodeTypeMutations(t *testing.T) {
 			fields: fields{
 				esVersion: "6.8.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masterdata-0").withVersion("6.8.0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
-					newTestPod("other-master-0").withVersion("6.8.0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masterdata-0").withVersion("6.8.0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("other-master-0").withVersion("6.8.0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchGreenHealth},
@@ -89,8 +89,8 @@ func TestUpgradePodsDeletion_WithNodeTypeMutations(t *testing.T) {
 			fields: fields{
 				esVersion: "7.2.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masterdata-0").withVersion("7.2.0").isMaster(true).isData(true).isHealthy(false).needsUpgrade(false).isInCluster(true),
-					newTestPod("other-master-0").withVersion("7.2.0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masterdata-0").withVersion("7.2.0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(false).needsUpgrade(false).isInCluster(true),
+					newTestPod("other-master-0").withVersion("7.2.0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 2, // 2 unavailable nodes to be sure that the predicate managing the masters is actually called
 				health:         client.Health{Status: esv1.ElasticsearchGreenHealth},
@@ -105,9 +105,9 @@ func TestUpgradePodsDeletion_WithNodeTypeMutations(t *testing.T) {
 			fields: fields{
 				esVersion: "7.2.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").withVersion("7.2.0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
-					newTestPod("data-to-masters-0").withVersion("7.2.0").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("data-to-masters-1").withVersion("7.2.0").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-0").withVersion("7.2.0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("data-to-masters-0").withVersion("7.2.0").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("data-to-masters-1").withVersion("7.2.0").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 2, // 2 unavailable nodes to be sure that the predicate managing the masters is actually called
 				health:         client.Health{Status: esv1.ElasticsearchGreenHealth},
@@ -122,9 +122,9 @@ func TestUpgradePodsDeletion_WithNodeTypeMutations(t *testing.T) {
 			fields: fields{
 				esVersion: "7.2.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").withVersion("7.2.0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
-					newTestPod("data-to-masters-0").withVersion("7.2.0").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("data-to-masters-1").withVersion("7.2.0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("masters-0").withVersion("7.2.0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("data-to-masters-0").withVersion("7.2.0").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("data-to-masters-1").withVersion("7.2.0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
 				),
 				maxUnavailable: 2, // 2 unavailable nodes to be sure that the predicate managing the masters is actually called
 				health:         client.Health{Status: esv1.ElasticsearchGreenHealth},
@@ -198,10 +198,10 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("master-0").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("node-0").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("node-1").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("node-2").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).isTerminating(true),
+					newTestPod("master-0").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("node-0").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("node-1").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("node-2").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).isTerminating(true),
 				),
 				maxUnavailable: 2,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -217,9 +217,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -235,9 +235,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -253,9 +253,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -271,9 +271,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 2,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -289,11 +289,11 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("master-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("master-1").isMaster(true).isData(true).isHealthy(false).needsUpgrade(true).isInCluster(false),
-					newTestPod("master-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("master-3").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("master-4").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("master-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("master-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(false).needsUpgrade(true).isInCluster(false),
+					newTestPod("master-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("master-3").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("master-4").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 2,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -309,8 +309,8 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("master-0").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("node-0").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("master-0").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("node-0").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -326,7 +326,7 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("master-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("master-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -342,7 +342,7 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("master-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("master-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -358,9 +358,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(false).needsUpgrade(true).isInCluster(false),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(false).needsUpgrade(true).isInCluster(false),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -376,8 +376,8 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(false).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(false).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth},
@@ -409,8 +409,8 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth},
@@ -441,7 +441,7 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth},
@@ -473,7 +473,7 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.5.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.5.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth},
@@ -505,9 +505,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.5.0"),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.5.0"),
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(false).needsUpgrade(true).isInCluster(false).withVersion("7.5.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.5.0"),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.5.0"),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(false).needsUpgrade(true).isInCluster(false).withVersion("7.5.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth},
@@ -538,9 +538,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth},
@@ -571,11 +571,11 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
-					newTestPod("masters-3").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-4").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
+					newTestPod("masters-3").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-4").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth},
@@ -620,9 +620,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.5.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.5.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth},
@@ -653,9 +653,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth, RelocatingShards: 1},
@@ -679,8 +679,8 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withVersion("7.4.0"),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true).withVersion("7.5.0"),
 				),
 				maxUnavailable: 1,
 				health:         client.Health{Status: esv1.ElasticsearchYellowHealth},
@@ -720,8 +720,8 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(false).needsUpgrade(true).isInCluster(false),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(false).needsUpgrade(true).isInCluster(false),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -737,9 +737,9 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-2").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -758,15 +758,15 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 				esVersion: "7.5.0",
 				upgradeTestPods: newUpgradeTestPods(
 					// 5 data nodes
-					newTestPod("elasticsearch-sample-es-nodes-4").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("elasticsearch-sample-es-nodes-3").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("elasticsearch-sample-es-nodes-2").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("elasticsearch-sample-es-nodes-1").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("elasticsearch-sample-es-nodes-0").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("elasticsearch-sample-es-nodes-4").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("elasticsearch-sample-es-nodes-3").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("elasticsearch-sample-es-nodes-2").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("elasticsearch-sample-es-nodes-1").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("elasticsearch-sample-es-nodes-0").withRoles(esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 					// 3 masters
-					newTestPod("elasticsearch-sample-es-masters-2").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("elasticsearch-sample-es-masters-1").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true),
-					newTestPod("elasticsearch-sample-es-masters-0").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("elasticsearch-sample-es-masters-2").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("elasticsearch-sample-es-masters-1").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("elasticsearch-sample-es-masters-0").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				shardLister:    migration.NewFakeShardFromFile("shards.json"),
 				maxUnavailable: 2, // Allow 2 to be upgraded at the same time
@@ -785,7 +785,7 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.15.2",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				shutdowns: map[string]client.NodeShutdown{
 					"masters-0": {Status: client.ShutdownStarted},
@@ -804,8 +804,8 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.15.2",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(false),
+					newTestPod("masters-1").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(false),
 				),
 				maxUnavailable: 1,
 				shardLister:    migration.NewFakeShardLister(client.Shards{}),
@@ -821,7 +821,7 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 			fields: fields{
 				esVersion: "7.15.2",
 				upgradeTestPods: newUpgradeTestPods(
-					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole, esv1.DataRole).isHealthy(true).needsUpgrade(true).isInCluster(true),
 				),
 				shutdowns: map[string]client.NodeShutdown{
 					"masters-0": {Status: client.ShutdownComplete},
@@ -887,12 +887,12 @@ func TestDeletionStrategy_SortFunction(t *testing.T) {
 			fields: fields{
 				upgradeTestPods: newUpgradeTestPods(
 					// use "amasters" rather than "masters" to ensure we are not relying on the name sort accidentally
-					newTestPod("amasters-0").isMaster(true).needsUpgrade(true),
-					newTestPod("data-0").isData(true).needsUpgrade(true),
-					newTestPod("masters-0").isMaster(true).needsUpgrade(true),
-					newTestPod("amasters-1").isMaster(true).needsUpgrade(true),
-					newTestPod("data-1").isData(true).needsUpgrade(true),
-					newTestPod("amasters-2").isMaster(true).needsUpgrade(true),
+					newTestPod("amasters-0").withRoles(esv1.MasterRole).needsUpgrade(true),
+					newTestPod("data-0").withRoles(esv1.DataRole).needsUpgrade(true),
+					newTestPod("masters-0").withRoles(esv1.MasterRole).needsUpgrade(true),
+					newTestPod("amasters-1").withRoles(esv1.MasterRole).needsUpgrade(true),
+					newTestPod("data-1").withRoles(esv1.DataRole).needsUpgrade(true),
+					newTestPod("amasters-2").withRoles(esv1.MasterRole).needsUpgrade(true),
 				),
 				esState: &testESState{
 					inCluster: []string{"data-1", "data-0", "amasters-2", "amasters-1", "amasters-0", "masters-0"},
@@ -906,12 +906,12 @@ func TestDeletionStrategy_SortFunction(t *testing.T) {
 			fields: fields{
 				upgradeTestPods: newUpgradeTestPods(
 					// use "amasters" rather than "masters" to ensure we are not relying on the name sort accidentally
-					newTestPod("amasters-0").isMaster(true).needsUpgrade(true),
-					newTestPod("amasters-1").isMaster(true).needsUpgrade(true),
-					newTestPod("amasters-2").isMaster(true).needsUpgrade(true),
-					newTestPod("data-0").isData(true).needsUpgrade(true),
-					newTestPod("data-1").isData(true).needsUpgrade(true),
-					newTestPod("data-2").isData(true).needsUpgrade(true),
+					newTestPod("amasters-0").withRoles(esv1.MasterRole).needsUpgrade(true),
+					newTestPod("amasters-1").withRoles(esv1.MasterRole).needsUpgrade(true),
+					newTestPod("amasters-2").withRoles(esv1.MasterRole).needsUpgrade(true),
+					newTestPod("data-0").withRoles(esv1.DataRole).needsUpgrade(true),
+					newTestPod("data-1").withRoles(esv1.DataRole).needsUpgrade(true),
+					newTestPod("data-2").withRoles(esv1.DataRole).needsUpgrade(true),
 				),
 				esState: &testESState{
 					inCluster: []string{"data-2", "data-1", "data-0", "amasters-2", "amasters-1", "amasters-0"},

--- a/pkg/controller/elasticsearch/driver/upgrade_predicates_test.go
+++ b/pkg/controller/elasticsearch/driver/upgrade_predicates_test.go
@@ -194,6 +194,39 @@ func TestUpgradePodsDeletion_Delete(t *testing.T) {
 		wantShardsAllocationDisabled bool
 	}{
 		{
+			name: "1 master and 2 data_content nodes, wait for all the data_content nodes to be upgraded first",
+			fields: fields{
+				esVersion: "7.15.0",
+				upgradeTestPods: newUpgradeTestPods(
+					newTestPod("master-0").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).withVersion("7.14.0").isInCluster(true),
+					newTestPod("node-0").withRoles(esv1.DataContentRole).isHealthy(true).needsUpgrade(true).withVersion("7.14.0").isInCluster(true),
+					newTestPod("node-1").withRoles(esv1.DataContentRole).isHealthy(true).needsUpgrade(true).withVersion("7.14.0").isInCluster(true),
+				),
+				maxUnavailable: 2,
+				shardLister: migration.NewFakeShardLister(client.Shards{
+					client.Shard{
+						Index:    "index_a",
+						Shard:    "0",
+						State:    "STARTED",
+						NodeName: "node-0",
+						Type:     "p",
+					},
+					client.Shard{
+						Index:    "index_a",
+						Shard:    "0",
+						State:    "STARTED",
+						NodeName: "node-1",
+						Type:     "r",
+					},
+				}),
+				health:    client.Health{Status: esv1.ElasticsearchGreenHealth},
+				podFilter: nothing,
+			},
+			deleted:                      []string{"node-1"},
+			wantErr:                      false,
+			wantShardsAllocationDisabled: true,
+		},
+		{
 			name: "Do not attempt to delete an already terminating Pod",
 			fields: fields{
 				esVersion: "7.5.0",

--- a/pkg/controller/elasticsearch/driver/upgrade_test.go
+++ b/pkg/controller/elasticsearch/driver/upgrade_test.go
@@ -152,9 +152,9 @@ func Test_healthyPods(t *testing.T) {
 			name: "All Pods are healthy",
 			args: args{
 				pods: newUpgradeTestPods(
-					newTestPod("masters-2").inStatefulset("masters").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
-					newTestPod("masters-1").inStatefulset("masters").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
-					newTestPod("masters-0").inStatefulset("masters").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
+					newTestPod("masters-2").inStatefulset("masters").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
+					newTestPod("masters-1").inStatefulset("masters").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
+					newTestPod("masters-0").inStatefulset("masters").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
 				),
 				statefulSets: sset.StatefulSetList{
 					sset.TestSset{
@@ -169,9 +169,9 @@ func Test_healthyPods(t *testing.T) {
 			name: "One Pod is terminating",
 			args: args{
 				pods: newUpgradeTestPods(
-					newTestPod("masters-2").inStatefulset("masters").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
-					newTestPod("masters-1").inStatefulset("masters").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true).isTerminating(true).withResourceVersion("999"),
-					newTestPod("masters-0").inStatefulset("masters").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
+					newTestPod("masters-2").inStatefulset("masters").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
+					newTestPod("masters-1").inStatefulset("masters").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true).isTerminating(true).withResourceVersion("999"),
+					newTestPod("masters-0").inStatefulset("masters").withRoles(esv1.MasterRole).isHealthy(true).needsUpgrade(true).isInCluster(true).withResourceVersion("999"),
 				),
 				statefulSets: sset.StatefulSetList{
 					sset.TestSset{

--- a/pkg/controller/elasticsearch/driver/upscale_test.go
+++ b/pkg/controller/elasticsearch/driver/upscale_test.go
@@ -417,9 +417,9 @@ func Test_adjustZenConfig(t *testing.T) {
 			es:          bootstrappedES,
 			statefulSet: sset.TestSset{Name: "masters", Version: "7.2.0", Replicas: 3, Master: true, Data: true},
 			pods: []runtime.Object{
-				newTestPod("masters-0").withVersion("6.8.0").isMaster(true).isData(true).toPodPtr(),
-				newTestPod("masters-1").withVersion("6.8.0").isMaster(true).isData(true).toPodPtr(),
-				newTestPod("masters-2").withVersion("6.8.0").isMaster(true).isData(true).toPodPtr(),
+				newTestPod("masters-0").withVersion("6.8.0").withRoles(esv1.MasterRole, esv1.DataRole).toPodPtr(),
+				newTestPod("masters-1").withVersion("6.8.0").withRoles(esv1.MasterRole, esv1.DataRole).toPodPtr(),
+				newTestPod("masters-2").withVersion("6.8.0").withRoles(esv1.MasterRole, esv1.DataRole).toPodPtr(),
 			},
 			wantMinimumMasterNodesSet: true,
 			wantInitialMasterNodesSet: false,

